roxctl/common/k8s_port_forward.go

```diff
@@ -0,0 +1,163 @@
+	"k8s.io/client-go/transport/spdy"
+	"k8s.io/kubectl/pkg/polymorphichelpers"
+	"k8s.io/kubectl/pkg/util/podutils"
+)
+
+var (
+	portForwardOnce     sync.Once
+	forwardEndpoint     string
+	portForwardingError error
+
+	errPortForwarding = errox.ServerError.New("port-forwarding error")
+)
+
+// setKubernetesDefaults sets default values on the provided client config for accessing the
+// Kubernetes API or returns an error if any of the defaults are impossible or invalid.
+// TODO this isn't what we want.  Each clientset should be setting defaults as it sees fit.
+func setKubernetesDefaults(config *restclient.Config) error {
+	// TODO remove this hack.  This is allowing the GetOptions to be serialized.
+	config.GroupVersion = &schema.GroupVersion{Group: "", Version: "v1"}
+
+	if config.APIPath == "" {
+		config.APIPath = "/api"
+	}
+	if config.NegotiatedSerializer == nil {
+		// This codec factory ensures the resources are not converted. Therefore, resources
+		// will not be round-tripped through internal versions. Defaulting does not happen
+		// on the client.
+		config.NegotiatedSerializer = scheme.Codecs.WithoutConversion()
+	}
+	if err := restclient.SetKubernetesDefaults(config); err != nil {
+		return errPortForwarding.CausedBy(err)
+	}
+	return nil
+}
+
+func runPortForward() (uint16, error) {
+	getter := genericclioptions.NewConfigFlags(true)
+	c := getter.ToRawKubeConfigLoader()
+	restConfig, err := c.ClientConfig()
+	if err != nil {
+		return 0, errPortForwarding.CausedBy(err)
+	}
+	namespace, _, err := c.Namespace()
+	if err != nil {
+		return 0, errPortForwarding.CausedBy(err)
+	}
+
+	if err := setKubernetesDefaults(restConfig); err != nil {
+		return 0, errPortForwarding.CausedBy(err)
+	}
+	coreClient, err := corev1client.NewForConfig(restConfig)
+	if err != nil {
+		return 0, errPortForwarding.CausedBy(err)
+	}
+	podClient, err := kubernetes.NewForConfig(restConfig)
+	if err != nil {
+		return 0, errPortForwarding.CausedBy(err)
+	}
+
+	sortBy := func(pods []*corev1.Pod) sort.Interface { return sort.Reverse(podutils.ActivePods(pods)) }
+	forwardablePod, _, err := polymorphichelpers.GetFirstPod(coreClient, namespace, "app=central", 10*time.Second, sortBy)
+	if err != nil {
+		return 0, errPortForwarding.CausedBy(err)
+	}
+
+	pod, err := podClient.CoreV1().Pods(namespace).Get(context.TODO(), forwardablePod.Name, metav1.GetOptions{})
+	if err != nil {
+		return 0, errPortForwarding.CausedBy(err)
+	}
+	if pod.Status.Phase != corev1.PodRunning {
+		return 0, fmt.Errorf("unable to forward port because pod is not running. Current status=%v", pod.Status.Phase)
+	}
+
+	client, err := restclient.RESTClientFor(restConfig)
+	if err != nil {
+		return 0, errPortForwarding.CausedBy(err)
+	}
+
+	signals := make(chan os.Signal, 1)
+	signal.Notify(signals, os.Interrupt)
+	defer signal.Stop(signals)
+
+	stopChannel := make(chan struct{}, 1)
+	readyChannel := make(chan struct{})
+
+	go func() {
+		<-signals
+		if stopChannel != nil {
```

Comment:
Is it possible to be stopChannel be nil?

Reply:
Not really.